book/src/spec.md

```diff
@@ -0,0 +1,628 @@
+Physical stream specification
+=============================
+
+A Tydi physical stream carries a stream of elements, dimensionality information
+for said elements, and (optionally) user-defined transfer information from a
+source to a sink. The contents of these three groups and to some extent the way
+in which they are transferred are defined in a parameterized way, to provide
+the necessary flexibility for the higher levels of this specification, and to
+give hardware developers the freedom to optimize their designs. Based on these
+parameters, this specification defines which signals must exist on the
+interface, and what the significance of these signals is.
+
+Parameters
+----------
+
+A physical stream interface has five parameters, described in the following
+subsections.
+
+> Note that we define the parameters on the interfaces of the source and the
+> sink, not on the stream between them! This is because the complexity
+> parameter need not be the same on the source and sink. The others do.
+
+### Element content (E)
+
+This describes the physical layout of the elementary data being transferred by
+the stream. It is an emptyable, ordered set of named fields, where each field
+has a name and a bit count.
+
+> The element type can be empty to make a "null" stream. Such streams can still
+> be useful, as physical streams also carry metadata.
+
+The letter \\(E\\) is used as a shorthand for the total number of bits in the
+element; that is, the sum of the field bit count over all fields in the
+element.
+
+#### Field name
+
+The name of each field is a string consisting of letters, numbers, and/or
+underscores.
+
+The name cannot start or end with an underscore.
+
+The name cannot start with a digit.
+
+> It is illegal to start or end with an underscore or start with a number to
```

Comment:
Then what do you suggest a backend does when an identifier given by the user can't be expressed? Name mangling? Hashing? Replace with underscores and hope for no conflicts? Hope that all languages (and all related tooling) support raw identifiers? Generate wrong code? Add checks in the backends that don't support the identifiers and have the user deal with it (that is, not implementing the full spec)?

The above is a subset of the identifier regex for all languages I know, keywords aside. VHDL just happens to be the worst one I know, too. I also know zero languages that allow identifiers to start with digits. So either we spec no restrictions on the identifier regex at all (with all aforementioned problems postponed), or we do the subset of what programming languages usually use (as suggested), but allowing `[\w]+` is just another standard to add on the pile.

Reply:
I would be fine with making this the responsibility of the backend, but I also agree that picking a subset of identifiers to avoid these problems is sane.